mesonbuild/environment.py

```diff
@@ -331,32 +333,46 @@ def __init__(self, source_dir, build_dir, options):
                 self.create_new_coredata(options)
             else:
                 raise e
-        self.exe_wrapper = None
 
         self.machines = MachineInfos()
         # Will be fully initialized later using compilers later.
         self.machines.detect_build()
-        if self.coredata.cross_file:
-            self.cross_info = CrossBuildInfo(self.coredata.cross_file)
-            if 'exe_wrapper' in self.cross_info.config['binaries']:
-                from .dependencies import ExternalProgram
-                self.exe_wrapper = ExternalProgram.from_bin_list(
-                    self.cross_info.config['binaries'], 'exe_wrapper')
-            if 'host_machine' in self.cross_info.config:
-                self.machines.host = MachineInfo.from_literal(
-                    self.cross_info.config['host_machine'])
-            if 'target_machine' in self.cross_info.config:
-                self.machines.target = MachineInfo.from_literal(
-                    self.cross_info.config['target_machine'])
-        else:
-            self.cross_info = None
-        self.machines.default_missing()
 
-        if self.coredata.config_files:
-            self.config_info = coredata.ConfigData(
+        # Similar to coredata.compilers and build.compilers, but lower level in
+        # that there is no meta data, only names/paths.
+        self.binaries = PerMachineDefaultable()
+
+        # Misc other properties about each machine.
+        self.properties = PerMachine(Properties(), Properties(), Properties())
+
+        if self.coredata.config_files is not None:
+            config_info = coredata.ConfigData(
                 coredata.load_configs(self.coredata.config_files))
         else:
-            self.config_info = coredata.ConfigData()
+            config_info = coredata.ConfigData()
+        self.binaries.build = BinaryTable(config_info.binaries)
+
+        if self.coredata.cross_file is not None:
+            config = ConfigFile(self.coredata.cross_file)
+            for (key, prop) in config.get('properties', {}).items():
+                self.properties.host[key] = prop
+            self.binaries.host = BinaryTable(config.get('binaries', {}), False)
+            if 'host_machine' in config:
+                self.machines.host = MachineInfo.from_literal(config['host_machine'])
+            if 'target_machine' in config:
+                self.machines.target = MachineInfo.from_literal(config['target_machine'])
+
+        self.machines.default_missing()
+        self.binaries.default_missing()
+
+        if self.binaries.host.binaries is not None and \
+           'exe_wrapper' in self.binaries.host.binaries:
+            from .dependencies import ExternalProgram
```

Comment:
This was how it was done before, see the diff. I bet it is indeed done to avoid a cycle, yuck. I agree it should be fixed, but I rather fix after this; consolidating the way binaries are found will help with moving that elsewhere.

Reply:
that's fine then, we can fix it later.